x/incentive/types/accumulator.go

```diff
@@ -0,0 +1,91 @@
+package types
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// An Accumulator handles calculating and tracking global reward distributions.
+type Accumulator struct {
+	PreviousAccumulationTime time.Time
+	Indexes                  RewardIndexes
+}
+
+func NewAccumulator(previousAccrual time.Time, indexes RewardIndexes) *Accumulator {
+	return &Accumulator{
+		PreviousAccumulationTime: previousAccrual,
+		Indexes:                  indexes,
+	}
+}
+
+// Accumulate accrues rewards up to the current time.
+// It calculates new rewards and adds them to the reward indexes for the period from PreviousAccumulationTime to currentTime.
+// It stores the currentTime in PreviousAccumulationTime to be used for later accumulations.
+// Rewards are not accrued for times outside of the start and end times of a reward period.
+// If a period ends before currentTime, the PreviousAccrualTime is shortened to the end time. This allows accumulate to be called sequentially on consecutive reward periods.
+func (acc *Accumulator) Accumulate(period MultiRewardPeriod, rewardSourceTotal sdk.Dec, currentTime time.Time) {
```

Comment:
Added. I also added a brief overview of how rewards work to the spec. It would be good to document everything more clearly but I'll prioritise it to after code freeze.